tests/jballerina-unit-test/src/test/resources/test-src/query/query-expr-type-narrowing-negative.bal

```diff
@@ -0,0 +1,496 @@
+    _ = from IntOrStr i in data
+        where i is int
+        select i * 2;
+
+    string[] _ = from IntOrStr i in data
+        where i !is int
+        select i;
+
+    check from IntOrStr i in data
+        where i is int
+        do {
+            int _ = i;
+        };
+
+    check from IntOrStr i in data
+        where i !is int
+        do {
+            string _ = i;
+        };
+}
+
+function test2(IntOrStr[] data1, IntOrStr[] data2) {
+    int[] _ = from IntOrStr i in data1
+        from IntOrStr j in data2
+        where i is int && j is int
+        select i * j;
+
+    string[] _ = from IntOrStr i in data1
+            from IntOrStr j in data2
+            where i !is int && j !is int
+            select i + j;
+
+    check from IntOrStr i in data1
+        from IntOrStr j in data2
+        where i !is int && j !is int
+        do {
+            string _ = i + j;
+        };
+}
+
+function test3(IntStrOrBoolean[] data) {
+    string[] _ = from IntStrOrBoolean i in data
+        where i !is boolean && i !is int
+        select i;
+
+    string[] _ = from IntStrOrBoolean i in data
+        where i !is boolean
+        where i !is int
+        select i;
+
+    int[][] _ = from IntStrOrBoolean i in data
+        where i !is boolean
+        select from int ii in 1...3
+        where i is int
+        select i * ii;
+
+    check from var item in data
+            where item !is boolean && item !is int
+            do {
+                string _ = item;
+            };
+
+    check from var i in data
+        where i !is boolean
+        do {
+            int[] _ = from int ii in 1 ... 3
+                where i is int
+                select i * ii;
+        };
+}
+
+type A record {|
+    int i;
+|};
+
+type B record {|
+    string i;
+|};
+
+type C record {|
+    string j;
+|};
+
+type AorB A|B;
+
+type AorC A|C;
+
+type BorC B|C;
+
+function test4(AorB[] data1, AorC[] data2, BorC[] data3) returns error? {
+    _ = from AorB item in data1
+        where item is A
+        select aFn(item);
+
+    _ = from AorB item in data1
+        where item !is A
+        select bFn(item); //error incompatible types: expected 'B', found 'AorB'
+
+    check from AorB item in data1
+        where !(item is A)
+        do {
+            bFn(item); //error incompatible types: expected 'B', found 'AorB'
+        };
+
+    _ = from AorB item in data1
+        where item !is A
+        select bFn(item); //error incompatible types: expected 'B', found 'AorB'
+
+    _ = from AorC item in data2
+        where item is A
+        select aFn(item);
+
+    _ = from AorC item in data2
+        where item !is A
+        select cFn(item);
+
+    _ = from BorC item in data3
+        where item is B
+        select bFn(item);
+
+    _ = from BorC item in data3
+        where item !is B
+        select cFn(item);
+
+    check from BorC item in data3
+        where item is B
+        do {
+            bFn(item);
+        };
+
+    check from BorC item in data3
```

Comment:
For completeness shall we have a similar query-action test for `AorC item in data2` as well?